agent/local/state.go

```diff
@@ -485,6 +521,20 @@ func (l *State) UpdateCheck(id types.CheckID, status, output string) {
 		return
 	}
 
+	// If this is a check for an aliased service, then notify the waiters.
+	if aliases, ok := l.checkAliases[c.Check.ServiceID]; ok && len(aliases) > 0 {
+		for _, notifyCh := range aliases {
+			// Do not block. All notify channels should be buffered to at
+			// least 1 in which case not-blocking does not result in loss
+			// of data because a failed send means a notification is
+			// already queued.
```

Comment:
I spent a while describing a race condition with this logic since the state output isn't actually updated at this point, but then realised it's all moot since this whole method is holding the state lock.

Maybe we could add a note here stating that as a necessary condition for the invariant you claimed? If an alias could drain it's notification chan and load the local check state before the status and output are updated below then there is a possible race. It's only mitigated because they can't get a lock to load the state until we're done with this whole method.